Fallback to use naive diff driver if enable CONFIG_OVERLAY_FS_REDIRECT_DIR

coolljt0725 opened the pull request:
When use overlay2 as the graphdriver and the kernel enable
`CONFIG_OVERLAY_FS_REDIRECT_DIR=y`, rename a dir in lower layer
will has a xattr to redirct its dir to source dir. This make the
image layer unportable. This patch fallback to use naive diff driver
when kernel enable CONFIG_OVERLAY_FS_REDIRECT_DIR

Signed-off-by: Lei Jitang <leijitang@huawei.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/docker/docker/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

coolljt0725 commented:
another approach for https://github.com/moby/moby/pull/34320

ping @dmcgowan  @rhvgoyal

AkihiroSuda reviewed (commented):
Typo: enbale

AkihiroSuda commented on daemon/graphdriver/overlay2/check.go:
```diff
@@ -8,17 +8,20 @@ import (
-func hasOpaqueCopyUpBug(d string) error {
+// directory or the kernel enable CONFIG_OVERLAY_FS_REDIRECT_DIR.
+// When these exist naive diff should be used.
+func hasOpaqueCopyUpBugOrEnbaleRedirectDir(d string) error {
```
still typo 😀

coolljt0725 replied:
fixed

rhvgoyal commented on daemon/graphdriver/overlay2/check.go:
```diff
@@ -8,17 +8,20 @@ import (
-func hasOpaqueCopyUpBug(d string) error {
+// directory or the kernel enable CONFIG_OVERLAY_FS_REDIRECT_DIR.
+// When these exist naive diff should be used.
+func hasOpaqueCopyUpBugOrEnableRedirectDir(d string) error {
```
Feels too long a name. May be new function should be in a separate helper function of its own instead of single function checking both.

coolljt0725 replied:
I just think there is no need to do another creating dir and mount/umount operations to check whether kernel enable redirect dir.

rhvgoyal replied:
I think that code is much more readable if we have it in separate function. Anyway it is called only once so overhead of creating directories and mounting should not matter much.

coolljt0725 replied:
I use a short func name instead, has it in separate function does make the code more readable, but it have some unnecessary  duplicate code and  overhead.

rhvgoyal commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -266,7 +266,7 @@ func supportsOverlay() error {
 func useNaiveDiff(home string) bool {
 	useNaiveDiffLock.Do(func() {
-		if err := hasOpaqueCopyUpBug(home); err != nil {
+		if err := hasOpaqueCopyUpBugOrEnableRedirectDir(home); err != nil {
```
So this is called on every invocation of Diff() and other places. I am wondering why it was not done once during initialization time and save the state in some boolean variable and use that for rest of the operations.

coolljt0725 replied:
This is only called once because  `useNaiveDiffLock` is a `sync.Once` lock

rhvgoyal replied:
Ok. I was not aware of `sync.Once`. So there must have been some reason that we could not do in Init() and had to rely on this sync.Once stuff.

coolljt0725 committed:
Fallback to use naive diff driver if enable CONFIG_OVERLAY_FS_REDIRECT_DIR

When use overlay2 as the graphdriver and the kernel enable
`CONFIG_OVERLAY_FS_REDIRECT_DIR=y`, rename a dir in lower layer
will has a xattr to redirct its dir to source dir. This make the
image layer unportable. This patch fallback to use naive diff driver
when kernel enable CONFIG_OVERLAY_FS_REDIRECT_DIR

Signed-off-by: Lei Jitang <leijitang@huawei.com>

coolljt0725 commented:
ping @dmcgowan   @stevvooe

rhvgoyal commented:
Ok, this patch looks fine to me.

LGTM

rhvgoyal commented:
cc @amir73il

rhvgoyal commented:
BTW, what does this change mean for performance. So now if we are running on a newer kernel which will most likely have `CONFIG_OVERLAY_FS_REDIRECT=y`. That means we will fallback to using naive diff interface which I am assuming is much slower as comapared to overlay2 native diff interface.

Does that mean, upgrading to a newer kernel effectively brings a performance regression?

rhvgoyal commented:
If CONFIG_OVERLAY_FS_REDIRECT is not a net win for docker, why not disable this feature at mount time and continue to use native diff. And provide an option for users to enable this feature.

In other words, default to disabling overlay fs redirect and allow users to opt in to enable this feature. That way we will not see performance regressions when newer kernels arrive. This assumes that diff is going to be a more frequent activity (atleast for build) as comapred to directory tree renames.

dmcgowan commented:
@rhvgoyal performance of diff will be slower, which effects build times. How significant that difference would require measurement. No matter what the performance difference is though, correctness of the layer diff is far more important here. It is also worth noting that when overlay2 (or overlay) is used for builds, there are still a few operations which will fail. Enabling this rename fixes one of those cases.

As an aside, I think it has been a bad idea to sacrifice possible correctness of the diff for performance. We had this issue with naive diff in the past not picking up differences in a small window because figuring out that diff may have required a more expensive hash calculation. For containerd 1.0 we have chosen a slightly different route of always using the equivalent of the naive diff and focusing on correctness, then only making optimizations on top of that algorithm which do not sacrifice correctness.

> If CONFIG_OVERLAY_FS_REDIRECT is not a net win for docker, why not disable this feature at mount time and continue to use native diff. And provide an option for users to enable this feature.

Having this as option could make sense. Right now this would have to be done in module and I don't believe we have updated any documentation for it. See related comment here https://github.com/moby/moby/issues/33733#issuecomment-309930404

LGTM

stevvooe commented on daemon/graphdriver/overlay2/check.go:
```diff
@@ -75,5 +81,23 @@ func hasOpaqueCopyUpBug(d string) error {
+	// rename "d1" to "d2"
+	if err := os.Rename(filepath.Join(td, "merged", "d1"), filepath.Join(td, "merged", "d2")); err != nil {
+		// if rename fail with syscall.EXDEV, the kernel doesn't enable CONFIG_OVERLAY_FS_REDIRECT_DIR
+		if strings.Contains(err.Error(), syscall.EXDEV.Error()) {
```
Any reason we can't check the error number directly? This can be downcast to an `(*os.LinkError)` and you can check against `(*os.LinkError).Err`. See https://golang.org/src/os/file_unix.go#L19 to see how the error is packed.

stevvooe commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -266,7 +266,7 @@ func supportsOverlay() error {
 	useNaiveDiffLock.Do(func() {
-		if err := hasOpaqueCopyUpBug(home); err != nil {
+		if err := doesSupportNativeDiff(home); err != nil {
 			logrus.Warnf("Not using native diff for overlay2: %v", err)
```
This log message might want to provide some information about the impact.

coolljt0725 commented:
@stevvooe  updated

thaJeztah commented:
@dmcgowan should we add a [warning](https://github.com/moby/moby/blob/8af4db6f002ac907b6ef8610b237879dfcaa5b7a/daemon/info_unix.go#L19-L29) to `docker info` if `CONFIG_OVERLAY_FS_REDIRECT` is enabled?

dmcgowan commented:
@thaJeztah we already have an info field for `Native Overlay Diff`, I think that is the relevant information here

justincormack commented:
@dmcgowan there is a `redirect_dir=off` mount option we can use to disable this, as well as the module option. So we could give the user a choice of disabling for performance or enabling for correctness. Not that keen on too many options though. In terms of backporting this for older docker versions just disabling it may make sense.

justincormack commented:
Also the commit https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/commit/?id=ff0f962ca3c38239b299a70e7eea27abfbb979c3 notes that enabling this will not work if you later downgrade your kernel.

rhvgoyal commented:
@dmcgowan One can disable this feature per mount also. So `-o redirect_dir=off` should do it. How about adding a mount option knob so that users can enable/disable it. By default, defaults will be determined by kernel or how module was loaded. And user can override that default using `--storage-opt overlay2.mountopt=redirect_dir=off/on`.

AkihiroSuda commented:
What's current status?
This seems to have high priority

thaJeztah commented:
ping @stevvooe @justincormack could you have another look if you have time? I see there's two non-maintainer LGTM's from @dmcgowan and @rhvgoyal that I'm happy to accept, but I'm not sure if there's anything left to do based on the last few comments

stevvooe commented:
@thaJeztah Does https://github.com/moby/moby/pull/34342#issuecomment-320669900 invalidate the need for this patch?

rhvgoyal commented:
@stevvooe I don't think that comment invalidates the need for this patch. So trying to think aloud again.

We have few issues at hand.

- If overlayfs has redirect_dir feature enabled, then generated image is unportable. This issue has to be solved anyway. And solution to switch to NaiveDiff() interface in that case sounds reasonable.

- Second issue is, what should be docker's default. Should docker enable redirect_dir by default or disable it by default or do nothing and just rely on what kernel has configured. I think we don't have a good answer for this yet. So for now, may be we can just rely on kernel default and not force the situation one way or other at all.

- Third issue is providing user a daemon option so that user can enable/disable this feature for their needs. Something like `--storage-opt overlay2.mountopt redirect_dir=on/off`. 

I think this patch addresses first issue but does not provide a knob for third issue. 

May be for now, we can merge this patch as it is. That way if kernel has redirect_dir enabled, docker will not be broken.

And in a later PR, we can provide a knob to enable/disable this feature at docker level.

dmcgowan commented on daemon/graphdriver/overlay2/overlay.go:
```diff
@@ -266,8 +266,8 @@ func supportsOverlay() error {
-		if err := hasOpaqueCopyUpBug(home); err != nil {
-			logrus.Warnf("Not using native diff for overlay2: %v", err)
+		if err := doesSupportNativeDiff(home); err != nil {
+			logrus.Warnf("Not using native diff for overlay2, this may brings a performance regression on building images: %v", err)
```
`this may cause degraded performance for building images`

dmcgowan commented:
One nit on the wording then LGTM

@rhvgoyal your assessment looks correct. As I was testing this on my local machine I noticed a few more xattrs that kind of surprised me as it was not documented. `trusted.overlay.origin` accompanies  `trusted.overlay.redirect`, but that is fine in this case since we are using naive diff. When `redirect_dir=off`, I noticed the opaque flag was there, this makes sense since it copied up the files and nothing underneath should be visible. However I also noticed a `trusted.overlay.impure` flag which is something we aren't currently handling. I am not convinced we can or should even try to keep up with these xattrs and maybe should just always do naive diff. We should accept this change as is, but it does bring up some follow up work which we probably need. cc @amir73il

rhvgoyal commented:
Recently @amir73il has done work on `index` feature as well and we have another option `index=on/off` as well. I suspect that might cause issues as well, I am not sure though. I don't understand that well.

amir73il commented:
@dmcgowan I agree that this patch looks like the sane first step to take.
going forward, the way to handle portability of overlay layers correctly IMO
is to have a tool to pack/unpack layers.
The pack tool could add more info needed for porting.
The unpack tool could expand directives like redirect_dir into copy up of
directory trees if target platform does not support redirect_dir. (i.e. convert
directory move to recursive tree move, just like mv tool does with EXDEV).

In any case, overlayfs does NOT handle portability of the new features
(redirect_dir, index) and leaves the problem to the application to solve.
So if docker is interested in providing users with new overlayfs features
(directory rename, not breaking hardlinks on copy up), docker should also
come up with a strategy for portable images with advanced features.

The basic strategy that many file systems take is to store feature sets
with the image and compare those with the supported features of the platform when
mounting the file system (compat,rocompat,incompat features).
We can discuss this later on a dedicated thread.

W.R.T your concerns about the new overlay xattrs that you saw,
trusted.overlay.origin and trusted.overlay.impure have to do with constant
inode number across copy up. trusted.overlay.nlink has to do with copying up
hardlinks. All those xattrs MUST NOT be copied to image.
Regarding index=on, as far as I can remember, there is no portability issue with
enabling index, so long as trusted.overlay.origin xattrs are NOT copied to image.

coolljt0725 commented:
@dmcgowan  updated per your suggestion

dmcgowan reviewed (approved):
LGTM

kolyshkin commented on daemon/graphdriver/overlay2/check.go:
```diff
@@ -75,5 +80,23 @@ func hasOpaqueCopyUpBug(d string) error {
 
+	// rename "d1" to "d2"
+	if err := os.Rename(filepath.Join(td, "merged", "d1"), filepath.Join(td, "merged", "d2")); err != nil {
+		// if rename fail with syscall.EXDEV, the kernel doesn't enable CONFIG_OVERLAY_FS_REDIRECT_DIR
```
Please change the comment to
```go
// if rename failed with syscall.EXDEV, the kernel doesn't have CONFIG_OVERLAY_FS_REDIRECT_DIR enabled
```

(as it's not the kernel which enables or disables its features, but someone who compiled this kernel).

kolyshkin commented on daemon/graphdriver/overlay2/check.go:
```diff
@@ -75,5 +80,23 @@ func hasOpaqueCopyUpBug(d string) error {
+	}
+
+	if string(xattrRedirect) == "d1" {
+		return errors.New("kernel has enabled CONFIG_OVERLAY_FS_REDIRECT_DIR")
```
please change to

```go
		return errors.New("kernel has CONFIG_OVERLAY_FS_REDIRECT_DIR enabled")
```

kolyshkin reviewed (changes requested):
Looks good to me except for some minor fixes in comments and error messages, please take a look

coolljt0725 commented:
@kolyshkin  updated

kolyshkin commented:
LGTM

vieux commented:
@tonistiigi do you mind taking a look at this one please ?

kolyshkin commented:
* s390z (aka z) test failure is known issue (https://github.com/moby/moby/issues/34938) and is unrelated

tonistiigi reviewed (approved):
LGTM

tonistiigi commented on daemon/graphdriver/overlay2/check.go:
```diff
@@ -75,5 +80,23 @@ func hasOpaqueCopyUpBug(d string) error {
+	}
+	// get the xattr of "d2"
+	xattrRedirect, err := system.Lgetxattr(filepath.Join(td, "l3", "d2"), "trusted.overlay.redirect")
+	if err != nil {
```
nit: comment here that this function returns nil for missing xattr would be nice as sys/os pkg, or for example https://github.com/opencontainers/runc/blob/master/libcontainer/system/xattrs_linux.go#L7 do not ignore error in this case.

kolyshkin commented:
CI on power is still running
https://jenkins.dockerproject.org/job/Docker-PRs-powerpc/6271/console
 (github failed to update its status, guess it will do so once it's done)

thaJeztah commented:
Hm, failure on `z` (probably not related, so restarting); 

```
23:41:33 --- FAIL: TestStateTimeoutWait (0.35s)
23:41:33 	state_test.go:137: Stop callback doesn't fire in 200 milliseconds
23:41:33 FAIL
23:41:33 coverage: 27.6% of statements
```

thaJeztah reviewed (approved):
I see quite some LGTM's here, so moving to "merge"

yongtang commented:
z failure is on #34938, merging

khimaros commented:
FYI, there seems to be a pretty major bug with this implementation #37965 and https://stackoverflow.com/questions/46787983/what-native-overlay-diff-mean-in-overlay2-storage-driver/49615244#49615244 for details.

oldthreefeng commented:
> @dmcgowan should we add a [warning](https://github.com/moby/moby/blob/8af4db6f002ac907b6ef8610b237879dfcaa5b7a/daemon/info_unix.go#L19-L29) to `docker info` if `CONFIG_OVERLAY_FS_REDIRECT` is enabled?

i this is very important info for user .   when this `CONFIG_OVERLAY_FS_REDIRECT`  is enabled . docker build will failed  and cause layer not found error  https://github.com/moby/moby/issues/37965  .   @thaJeztah
